### PR TITLE
Documentation for MTF linters

### PR DESCRIPTION
**Pull request by phracek**
Signed-off-by: Petr "Stone" Hracek <phracek@redhat.com>

This Pull Request introduces docuementation about MTF linters.
Reference to MTF issue is: #202

It contains as dockerfile linters as help file linters.

### CODE REVIEW

File: docs/user_guide/mtf_linters.rst [hunk 1]
```diff
@@ -1,8 +1,70 @@
 Linters
 =================
 
-MTF provides a set of linters for checking containers
+MTF provides a set of linters for checking containers and help files
```
Comment (TomasTomecek) [inline docs/user_guide/mtf_linters.rst]: and dockerfiles

File: docs/user_guide/mtf_linters.rst [hunk 2]
```diff
@@ -1,8 +1,70 @@
...
 
 Dockerfile linters
 ~~~~~~~~~~~~~~~~~~
+Dockerfile linters are divided into two python files, like `dockerlint.py`_ and `dockerfile_lint.py`_.
```
Comment (TomasTomecek) [inline docs/user_guide/mtf_linters.rst]: it's perfectly fine to just type: *two python files: `dock...*

File: docs/user_guide/mtf_linters.rst [hunk 3]
```diff
@@ -1,8 +1,70 @@
...
+
+.. _dockerlint.py: https://github.com/fedora-modularity/meta-test-family/blob/master/moduleframework/tests/generic/dockerlint.py
+.. _dockerfile_lint.py: https://github.com/fedora-modularity/meta-test-family/blob/master/moduleframework/tests/static/dockerfile_lint.py
+
+`dockerlint.py`_ contains those checks inside built image:
```
Comment (TomasTomecek) [inline docs/user_guide/mtf_linters.rst]: *performs these checks on an existing container image:*

File: docs/user_guide/mtf_linters.rst [hunk 4]
```diff
@@ -1,8 +1,70 @@
...
+
+* **test_all_nodocs** checks if documentation files exist in image. There are installed in the ``base image`` or by ``RUN`` commands." . This is only ``WARN`` check.
```
Comment (TomasTomecek) [inline docs/user_guide/mtf_linters.rst]: *files shipped by an installed RPM package*

File: docs/user_guide/mtf_linters.rst [hunk 5]
```diff
@@ -1,8 +1,70 @@
...
+* **test_installed_docs** checks if packages installed by ``RUN dnf`` command installs also documentation files.
```
Comment (TomasTomecek) [inline docs/user_guide/mtf_linters.rst]: ehm, isn't this description meant to be for the first item?

File: docs/user_guide/mtf_linters.rst [hunk 6]
```diff
@@ -1,8 +1,70 @@
...
+* **test_clean_all** checks if ``dnf/yum clean all`` is present in Dockerfile.
+
+`dockerfile_lint.py`_ contains those checks in Dockerfile file:
```
Comment (TomasTomecek) [inline docs/user_guide/mtf_linters.rst]: *these checks are performed on a Dockerfile*

File: docs/user_guide/mtf_linters.rst (hunk 5 above)
Comment (phracek) [reply docs/user_guide/mtf_linters.rst]: no, this check only checks doc files except base_image.

**Comment by TomasTomecek** [comment]
I can still see some grammar deficiencies in some sentences and I would love if someone else could take a look.

File: docs/user_guide/mtf_linters.rst [hunk 7]
```diff
@@ -1,8 +1,70 @@
 Linters
 =================
 
-MTF provides a set of linters for checking containers
+MTF provides a set of linters for checking containers, help files and dockerfiles.
```
Comment (eliskasl) [inline docs/user_guide/mtf_linters.rst]: s/dockerfiles/Dockerfiles/

File: docs/user_guide/mtf_linters.rst [hunk 8]
```diff
@@ -1,8 +1,70 @@
...
 Dockerfile linters
 ~~~~~~~~~~~~~~~~~~
+Dockerfile linters are divided into two python files: `dockerlint.py`_ and `dockerfile_lint.py`_.
+
+.. _dockerlint.py: https://github.com/fedora-modularity/meta-test-family/blob/master/moduleframework/tests/generic/dockerlint.py
+.. _dockerfile_lint.py: https://github.com/fedora-modularity/meta-test-family/blob/master/moduleframework/tests/static/dockerfile_lint.py
+
+`dockerlint.py`_ performs these checks on an existing container image:
+
+* **test_all_nodocs** checks if documentation files shipped by an install RPM packages. There are installed in the ``base image`` or by ``RUN`` commands." . This is only ``WARN`` check.
```
Comment (eliskasl) [inline docs/user_guide/mtf_linters.rst]: I'm only assuming what this option does, so correct me if I'm wrong. How about: ...checks if documentations files shipped by installed RPM packages have been removed. They are usually installed in the ``base image`` and inherited by child layer or installed via the ``RUN`` instruction. ...

File: docs/user_guide/mtf_linters.rst [hunk 9]
```diff
@@ -1,8 +1,70 @@
...
+* **test_installed_docs** checks if RPM packages installed by ``RUN dnf`` command installs also documentation files. Except ``base image``.
```
Comment (eliskasl) [inline docs/user_guide/mtf_linters.rst]: s/by ``RUN dnf`` command installs also/by the ``RUN dnf`` command also installs/

s/Except base image/The base image is an exception./

File: docs/user_guide/mtf_linters.rst [hunk 10]
```diff
@@ -1,8 +1,70 @@
...
+* **test_clean_all** checks if ``dnf/yum clean all`` is present in Dockerfile.
+
+`dockerfile_lint.py`_ these checks are performed on a Dockerfile:
+
+* **test_from_is_first_directive** checks if FROM instruction is really first in Dockerfile.
```
Comment (eliskasl) [inline docs/user_guide/mtf_linters.rst]: the FROM

File: docs/user_guide/mtf_linters.rst [hunk 11]
```diff
@@ -1,8 +1,70 @@
...
+* **test_from_directive_is_valid** checks if FROM instruction has proper format.
```
Comment (eliskasl) [inline docs/user_guide/mtf_linters.rst]: the FROM

File: docs/user_guide/mtf_linters.rst [hunk 12]
```diff
@@ -1,8 +1,70 @@
...
+* **test_chained_run_dnf_commands** checks if dnf/yum commands are chained or not.
```
Comment (eliskasl) [inline docs/user_guide/mtf_linters.rst]: if THE dnf/yum

File: docs/user_guide/mtf_linters.rst [hunk 13]
```diff
@@ -1,8 +1,70 @@
...
+* **test_checked_run_rest_commands** checks if RUN instructions, except dnf/yum, are chained or not.
+* **test_helpmd_is_present** checks if help file is present for this container.
```
Comment (eliskasl) [inline docs/user_guide/mtf_linters.rst]: THE help...

File: docs/user_guide/mtf_linters.rst [hunk 14]
```diff
@@ -1,8 +1,70 @@
...
+* **test_architecture_label_exists** checks if architecture label is present in Dockerfile.
```
Comment (eliskasl) [inline docs/user_guide/mtf_linters.rst]: if THE architecture...

File: docs/user_guide/mtf_linters.rst [hunk 15]
```diff
@@ -1,8 +1,70 @@
...
+* **test_name_in_env_and_label_exists** checks if name label is present in Dockerfile and NAME is present as ENV variable.
```
Comment (eliskasl) [inline docs/user_guide/mtf_linters.rst]: if THE name ... as an EVN

File: docs/user_guide/mtf_linters.rst [hunk 16]
```diff
@@ -1,8 +1,70 @@
...
+* **test_maintainer_label_exists** checks if maintainer label is present in Dockerfile.
```
Comment (eliskasl) [inline docs/user_guide/mtf_linters.rst]: if THE :)

File: docs/user_guide/mtf_linters.rst [hunk 17]
```diff
@@ -1,8 +1,70 @@
...
+* **test_release_label_exists** checks if release label is present in Dockerfile.
```
Comment (eliskasl) [inline docs/user_guide/mtf_linters.rst]: if THE release :)

File: docs/user_guide/mtf_linters.rst [hunk 18]
```diff
@@ -1,8 +1,70 @@
...
+* **test_version_label_exists** checks if version label is present in Dockerfile.
```
Comment (eliskasl) [inline docs/user_guide/mtf_linters.rst]: if THE :)

File: docs/user_guide/mtf_linters.rst [hunk 19]
```diff
@@ -1,8 +1,70 @@
...
+* **test_com_redhat_component_label_exists** checks if com.redhat.component label is present in Dockerfile.
```
Comment (eliskasl) [inline docs/user_guide/mtf_linters.rst]: if THE... :)

File: docs/user_guide/mtf_linters.rst [hunk 20]
```diff
@@ -1,8 +1,70 @@
...
+* **test_summary_label_exists** checks if summary label is present in Dockerfile.
```
Comment (eliskasl) [inline docs/user_guide/mtf_linters.rst]: if THE :)

File: docs/user_guide/mtf_linters.rst [hunk 21]
```diff
@@ -1,8 +1,70 @@
...
+* **test_run_or_usage_label_exists** check is run or usage label is present in Dockerfile.
```
Comment (eliskasl) [inline docs/user_guide/mtf_linters.rst]: s/check is run/check if the run/

File: docs/user_guide/mtf_linters.rst [hunk 22]
```diff
@@ -1,8 +1,70 @@
...
+
+Help file linter
+~~~~~~~~~~~~~~~~~
+Help file linter checks if help.md file contains important sections. Help file linter is `helpmd_lint.py`_.
```
Comment (eliskasl) [inline docs/user_guide/mtf_linters.rst]: if THE help.md

File: docs/user_guide/mtf_linters.rst [hunk 23]
```diff
@@ -1,8 +1,70 @@
...
+
+`helpmd_lint.py`_ contains those checks inside help.md file:
+
+* **test_helpmd_image_name** checks if help file contains image name. The correct format is e.g. ``% MEMCACHED(1)``.
+* **test_helpmd_maintainer_name** checks if help file contains maintainer name. The correct format is e.g. ``% USER NAME``.
+* **test_helpmd_name** checks if help file contains section called ``# NAME``. The section describes name of the container and short description.
+* **test_helpmd_description** checks if help file contains section called ``# DESCRIPTION``. This sections describes how to use image, etc.
+* **test_helpmd_usage** checks if help file contains section called ``# USAGE``.
+* **test_helpmd_environment_variables** checks if help file contains section called ``# ENVIRONMENT VARIABLES``. The check is valid only if ENV variable are present in Dockerfile. There is no heuristic check, like validation all variables.
+* **test_helpmd_security_implications** checks if help file contains section called ``# SECURITY IMPLICATIONS``. The check is valid only if container expose a port. There is no heuristic if exposed port is the same in help file.
```
Comment (eliskasl) [inline docs/user_guide/mtf_linters.rst]: missing some articles just like pointed above. Plus:
The check is valid only if container exposeS a port. There is no heuristic if the exposed port is the same as specified in the help file

File: docs/user_guide/mtf_linters.rst [hunk 24]
```diff
@@ -1,8 +1,70 @@
...
+   # USAGE
+   # SECURITY IMPLICATIONS
+
+
+`helpmd_lint.py`_ contains those checks inside help.md file:
+
+* **test_helpmd_image_name** checks if help file contains image name. The correct format is e.g. ``% MEMCACHED(1)``.
+* **test_helpmd_maintainer_name** checks if help file contains maintainer name. The correct format is e.g. ``% USER NAME``.
+* **test_helpmd_name** checks if help file contains section called ``# NAME``. The section describes name of the container and short description.
+* **test_helpmd_description** checks if help file contains section called ``# DESCRIPTION``. This sections describes how to use image, etc.
```
Comment (eliskasl) [inline docs/user_guide/mtf_linters.rst]: if THE help file contains a section...

File: docs/user_guide/mtf_linters.rst (hunk 8 above)
Comment (phracek) [reply docs/user_guide/mtf_linters.rst]: yes

Comment (TomasTomecek) [reply docs/user_guide/mtf_linters.rst]: *`documentation files`?

File: docs/user_guide/mtf_linters.rst [hunk 25]
```diff
@@ -1,8 +1,70 @@
...
 ~~~~~~~~~~~~~~~~~~
+Dockerfile linters are divided into two python files: `dockerlint.py`_ and `dockerfile_lint.py`_.
+
+.. _dockerlint.py: https://github.com/fedora-modularity/meta-test-family/blob/master/moduleframework/tests/generic/dockerlint.py
+.. _dockerfile_lint.py: https://github.com/fedora-modularity/meta-test-family/blob/master/moduleframework/tests/static/dockerfile_lint.py
+
+`dockerlint.py`_ performs these checks on an existing container image:
+
+* **test_all_nodocs** checks if documentation files shipped by an install RPM packages. There are installed by the ``base image`` or by the ``RUN`` commands." . This is only ``WARN`` check.
+* **test_installed_docs** checks if RPM packages installed by the ``RUN dnf`` command also installs documentation files. The ``base image`` is an exception.
```
Comment (TomasTomecek) [inline docs/user_guide/mtf_linters.rst]: should be `command also install doc...`

File: docs/user_guide/mtf_linters.rst [hunk 26]
```diff
@@ -1,8 +1,70 @@
...
 Dockerfile linters
 ~~~~~~~~~~~~~~~~~~
+Dockerfile linters are divided into two python files: `dockerlint.py`_ and `dockerfile_lint.py`_.
+
+.. _dockerlint.py: https://github.com/fedora-modularity/meta-test-family/blob/master/moduleframework/tests/generic/dockerlint.py
+.. _dockerfile_lint.py: https://github.com/fedora-modularity/meta-test-family/blob/master/moduleframework/tests/static/dockerfile_lint.py
+
+`dockerlint.py`_ performs these checks on an existing container image:
+
+* **test_all_nodocs** checks if documentation files shipped by an install RPM packages. There are installed by the ``base image`` or by the ``RUN`` commands." . This is only ``WARN`` check.
```
Comment (TomasTomecek) [inline docs/user_guide/mtf_linters.rst]: Not addressed: https://github.com/fedora-modularity/meta-test-family/pull/213#discussion_r169404990

Comment (phracek) [reply docs/user_guide/mtf_linters.rst]: Fixed. I did not read the comment properly.

**Comment by eliskasl** [comment]
LGTM! =^..^=

**Pull request merged**